### PR TITLE
Optimize native build of coreclr repo - 3x less memory , 11% build time improvement 

### LINKED ISSUE #14406: [Infrastructure/Windows] Native coreclr build spawns cl instances up to square of processor/core number and consumes 14 GB on 8 core box
Not sure that it is the right repo to post issue. If not please advise and I will move it.

Testing source-build I have noticed that on my notebook with 16 GB of RAM and 4 core / 8 thread processor MSBuild spawns total of 8 MSBuild nodes and each node may spawn in turn 8 compiling MSVC cl.exe processess. Compilers are starved and utilize on average 1 - 2 % of CPU time each and momory consumption varies from 50 MB up to 500 MB for each cl instance. Memory consumption of MSBuild process tree grows up to 14 GB - maximum from 3 runs.

In contrast to that situation build on linux (Ubuntu 14.04 and 16.04) and macOS consumes 1 - 2 GB, spawns 8 clang processes, each process uses roughly 12 - 15 % of processor time and memory usage ranges from 50 to maximum 200 MB. Build seems to be significantly faster as well but since I can not compare build times directly (my Mac box has significantly different specs than Windows one on which I use linux VM/WSL) I would consider this judgment to be not very precise.

This problem seems to be common to many different projects relying on CMake generating Visual Studio projects i.e. LLVM/Clang suffers from the very same issue. Therefore imho it could be helpful to change MSBuild configuration for native coreclr build to use single node instead of multiple nodes each for one virtual processor present in the machine.  

Situations seems to have improved with regard to memory consumption in last couple of weeks for master branch  but it still is highly inefficient in release/2.0.0 branch. Anyway having 64 compilers running in parallel on 8 cores wastes a lot of resources and time for synchronisation and context switching.

### PR DESCRIPTION
**Pull request by 4creators**
Fixes #14406 

After this commit is accepted and merged I will provide PR with build docs update

Native build in coreclr repo suffers from overallocation of VC compiler instnces
    which is a result of buggy parallelism handling by MSBuild for vcxproj. The machanism
    is such that MSBuild allocates nodes (default number of nodes is equal to number
    of logical processors on machine) which are capable of handling single vcxproj
    from build graph each yet still every node invokes CL compiler passing to it
    /MP option with multiple files which should be compiled in a given project.

This triggers parallel compilation mechanism in CL and next group of CL subprocesses
    is spawned equal to number of nodes for each node. Consequently during native compilation
    we can see 64 CL processes on 4 core / 8 thread machine which can consume up to 14 GB
    of memory. Overallocation mechanism is exponential and as a result on 8 core / 16 thread
    machine each processor core will handle 16 CL processess with theoretically 2 times
    higher memory consumption per core - total memory usage would grow from 14 GB to 56 GB.

MSBuild does not provide any command line switches which could be used to directly
    solve this problem. Setting /m:x will affect both - number of nodes and number of
    CL processess spawned by each node which will be equal to number of nodes. There
    is no direct method to decouple number of MSBuild nodes and number of CL processess
    spawned by each node.

To achive this without changing MSBuild and CL code it is necessary to use workaround
    and set values independetly. It is done by using /p:ForceImportBeforeCppTargets MSBuild
    command line switch and setting /m:2 global maximum CPU switch. First switch forces
    import of clr.nativebuild.props file contining maximum CPU settings used by CL compiler
    and set to experimentally optimized number equal to number of enabled processor cores -
    what is rather an expected value. Additionally it bypassess setting of MSBuild parallel
    compilation general switch by setting MultiProcessorCompilation value to true.

MSBuild /m:2 switch works incorrectly with vcxproj builds since it spawns as expected
    only 2 nodes but the limit of top level CL processes allowed is 3 - spawning is done
    indirectly with Tracker.exe executable. This means that with /m:2 and maximum CL count
    being set to 4 we will get maximum 12 CL processes and not as expected 8 CL processess.

 After experimentally determining optimal /m:x and CL process count combinations it was
    found that on i7-4700MQ Intel processor with 16 GB RAM memory consumption could be reduced
    almost 3 times to 2.8 GB from 8 GB, and compilation time was reduced from 00:14:35 
    (average of 3 runs) to 00:12:58 (average of 3 runs) by 11% (00:01:39).

 It is possible to further reduce amount of memory used during compilation but it does
    negatively impact compilation time.

 It is possible to further optimize build process but this would require changes to
    MSBuild and to Tracker or eventually CL executables.

### CODE REVIEW

**Comment by 4creators** [comment]
PTAL
@weshaggard @danmosemsft @wtgodbe @jkotas

File: build.cmd [hunk 1]
```diff
@@ -485,7 +493,15 @@ if /i "%__DoCrossArchBuild%"=="1" (
     set __MsbuildWrn=/flp1:WarningsOnly;LogFile=!__BuildWrn!
     set __MsbuildErr=/flp2:ErrorsOnly;LogFile=!__BuildErr!
 
-    @call %__ProjectDir%\run.cmd build -Project=%__CrossCompIntermediatesDir%\install.vcxproj -configuration=%__BuildType% -platform=%__CrossArch% -MsBuildLog=!__MsbuildLog! -MsBuildWrn=!__MsbuildWrn! -MsBuildErr=!__MsbuildErr! %__RunArgs% %__UnprocessedBuildArgs%
+    if not defined NumberOfEnabledCore (
```
Comment (danmosemsft) [inline build.cmd]: you can make this a batch file subroutine (with `call`) to avoid repeating the code, or presumably just put it right at the top of the file so it always runs.

File: build.cmd [hunk 2]
```diff
@@ -485,7 +493,15 @@ if /i "%__DoCrossArchBuild%"=="1" (
...
+    REM Determine number of physical processor cores available on machine
+      for /f "tokens=*" %%I in (
+        'wmic cpu get NumberOfEnabledCore /value ^| find "=" 2^>NULL'
+      ) do set %%I
+    )
+    echo Number of available CPU cores %NumberOfEnabledCore%
```
Comment (danmosemsft) [inline build.cmd]: Physical cores... I assume you picked this over logical cores by measurements?

**Comment by danmosemsft** [comment]
@4creators it is great that you did this. How you did it seems like a reasonable way.

do you need a similar change in build-test.cmd ?

**Comment by danmosemsft** [comment]
> MSBuild /m:2 switch works incorrectly with vcxproj builds since it spawns as expected
only 2 nodes but the limit of top level CL processes allowed is 3 - spawning is done
indirectly with Tracker.exe executable. This means that with /m:2 and maximum CL count
being set to 4 we will get maximum 12 CL processes and not as expected 8 CL processess.

I believe all you are seeing here is the way that CL /MP works. If I recall correctly, CL launches one child for each concurrent compilation. So CL /MP2 leads to 1 parent and 2 child CL processes but the parent does nothing, just waits on the children. Tracker should not be relevant - it is just a pass through executable that launches the child process it is instructed to launch, but with file IO detoured. It knows nothing about CL or multiple processors.

**Comment by danmosemsft** [comment]
Do you have numbers from the other combinations you tried, out of curiosity?

File: build.cmd (hunk 1 above)
Comment (4creators) [reply build.cmd]: Will move it into top of the file

File: build.cmd (hunk 2 above)
Comment (4creators) [reply build.cmd]: Yes, this result of optimization runs, however to some extent it was expected as compilation is processor bound and most of HPC loads run best when hyperthreading is off and are optimized with regard to number of physical cores.

File: build.cmd (hunk 1 above)
Comment (mikedn) [reply build.cmd]: How is this different from using the predefined environment variable `NUMBER_OF_PROCESSORS` ?

**Comment by 4creators** [comment]
> do you need a similar change in build-test.cmd ?

@danmosemsft yes I think it is worth to take a look at this - I will do it once we get this PR in

> Do you have numbers from the other combinations you tried, out of curiosity?

Yes, memory usage for runs with setting 3 or 5 processors was similar but timing was worse - 13:39.33 for 3 processors and 14:37.58 for 5 processor setting. Below are numbers from release/2.0.0 branch which was most painful as it is using up to 14 GB during build via source-build repo. Memory usage is calculated as net difference between rest state and maximum during test run.

CoreCLR v2.0.2 commit d0fbcf46ab7b5c26297b36a4580a0f89b6a372e7

- Current build configuration

10 825 MB - 15:08.14
10 693 MB - 15:35.31
08 740 MB - 15:12.73

- Mod 6 processor per 3 leafs (leaf means number of MSBuild spawned Tracker nodes)

3 125 MB 14:40.50

- Mod 5 processors per 3 leafs

3 452 MB 13:20.28

- Mod 4 proc per 3 leafs

2 547 MB 13:06.74 
		12:43.80 
		12:44.63
2 853 MB	13:01.28

- Mod 3 proc per 3 leafs

2 202 MB 13:24.73
2 223 MB 13:23.24
2 200 MB 13:24.47

- Mod 3 proc per 5 leafs

3 250 GB 15:22.25

**Comment by 4creators** [comment]
> I believe all you are seeing here is the way that CL /MP works

@danmosemsft 

I thought so as well at the beginning but after a while I realized that there are some problems with this hypothesis. First I have tested build configuration with switches /m:8 and /p:BuildInParallel=false. This seems to be the best configuration in the case MSBuild would use one node and CL would spawn 8 CL compilers. And it did happen but with one restriction: MSBuild was sequentially traversing projects graph building one project after the other what often resulted in sub optimal number of CL processes and no parallelism during linking. Memory usage was best (just under 2 GB) but build time was worse than current configuration.

Than I have tested /m:2 and /m:3 switches which resulted in spawning 2 and 3 MSBuild nodes respectively but than each of MSBuild sub nodes was spawning two top level CL nodes (Tracker instances) while top level MSBuild node was spawning one top level CL node. This resulted in 3 and 5 top level CL nodes respectively. Due to this result I am sure there is a MSBuild bug which leads to creation of 2 CL nodes per MSBuild node instead of expected one. This could be related to how MSBuild handles  project build parallelism for native builds.

File: build.cmd (hunk 1 above)
Comment (4creators) [reply build.cmd]: WMI which is used here provides several data points on machine: `NumberOfProcessors` which is equal to environment `NUMBER_OF_PROCESSORS` provides number of logical processors which comprises all threads available on processors implementing SMT. WMI property `NumberOfCores` provides physical core count of the CPU, whereas `NumberOfEnabledCore` property provides count of enabled cores on the CPU (some cores can be disabled via BIOS). Optimization runs indicated that number on which parallelisation should be based is a number of cores and not a number of threads and since some cores can be off it is necessary to retrieve enabled core count via WMI.

File: build.cmd [hunk 3]
```diff
@@ -415,7 +425,7 @@ if %__BuildNative% EQU 1 (
     set __MsbuildWrn=/flp1:WarningsOnly;LogFile=!__BuildWrn!
     set __MsbuildErr=/flp2:ErrorsOnly;LogFile=!__BuildErr!
 
-    @call %__ProjectDir%\run.cmd build -Project=%__IntermediatesDir%\install.vcxproj -MsBuildLog=!__MsbuildLog! -MsBuildWrn=!__MsbuildWrn! -MsBuildErr=!__MsbuildErr! -configuration=%__BuildType% %nativePlatfromArgs% %__RunArgs% %__UnprocessedBuildArgs%
+    @call %__ProjectDir%\run.cmd build -Project=%__IntermediatesDir%\install.vcxproj -MsBuildLog=!__MsbuildLog! -MsBuildWrn=!__MsbuildWrn! -MsBuildErr=!__MsbuildErr! -configuration=%__BuildType% %nativePlatfromArgs% %__RunArgs% -ExtraParameters="/p:ForceImportBeforeCppTargets=%__ProjectDir%/clr.nativebuild.props /m:2" %__UnprocessedBuildArgs%
```
Comment (weshaggard) [inline build.cmd]: Any reason you are passing this as a parameter instead of just including clr.nativebuild.props via dir.props? Actually thinking about that I'm not entirely sure dir.props gets imported, but perhaps we can tell cmake to import it in one of the cmake files instead of having to remember to pass this extra property everywhere we do a root native build.

File: clr.nativebuild.props [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <CL_MPCount>$(NumberOfEnabledCore)</CL_MPCount>
```
Comment (weshaggard) [inline clr.nativebuild.props]: Unfortunately I don't know enough about these properties to say this is the best way or not. I will see if I can find a contact in C++ land to review this change as well. Although please do not block on that if this has improvements then lets do it.

File: clr.nativebuild.props [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+    <ProcessorNumber>$(CL_MPCount)</ProcessorNumber>
+    <MultiProcessorCompilation>true</MultiProcessorCompilation>
```
Comment (danmosemsft) [inline clr.nativebuild.props]: I'm not sure that $(MultiProcessorCompilation) is a property? In the C++ build I see this going to CL
`MultiProcessorCompilation="%(ClCompile.MultiProcessorCompilation)"`
that means that it is item metadata. It has to either be set on ClCompile items individually, or on all of them using ItemDefinitionGroup.

$(CL_MPCount) is what does the trick I think -- it is converted to metadata
```
<ProcessorNumber Condition="'$(CL_MPCount)'                              != ''">$(CL_MPCount)</ProcessorNumber>
...
<CL
ProcessorNumber="%(ClCompile.ProcessorNumber)"
```

and I guess that setting ProcessorNumber (/MP:n) implies MultiprocessorCompilation (/MP).

Sicne CL_MPCount sets ProcessorNumber, I think only CL_MPCount is required, of these three.

Comment (danmosemsft) [reply clr.nativebuild.props]: I got all this info by doing "msbuild /pp" on a vcxproj.

File: build.cmd (hunk 3 above)
Comment (4creators) [reply build.cmd]: Actually it is a solution which did work. I admit that I have not tried all possible inclusion points due to time burden of working in 15 min code - test cycle. If there is a simpler solution to get this properties set for all native projects why not use it. Unfortunately CMake is not an area where I feel very comfortable.

File: clr.nativebuild.props (hunk 5 above)
Comment (4creators) [reply clr.nativebuild.props]: Yeah, confirm CL_MPCount is required and sufficient to control number of CL instances. Setting MultiProcessorCompilation to false can be overridden by CL_MPCount > 1.

Comment (danmosemsft) [reply clr.nativebuild.props]: You could even pass it with /p globally instead of using this file. But using a file like this means that individual projects can modify the value themselves if they want.

File: build.cmd [hunk 6]
```diff
@@ -484,8 +494,8 @@ if /i "%__DoCrossArchBuild%"=="1" (
     set __MsbuildLog=/flp:Verbosity=normal;LogFile=!__BuildLog!
     set __MsbuildWrn=/flp1:WarningsOnly;LogFile=!__BuildWrn!
     set __MsbuildErr=/flp2:ErrorsOnly;LogFile=!__BuildErr!
-
-    @call %__ProjectDir%\run.cmd build -Project=%__CrossCompIntermediatesDir%\install.vcxproj -configuration=%__BuildType% -platform=%__CrossArch% -MsBuildLog=!__MsbuildLog! -MsBuildWrn=!__MsbuildWrn! -MsBuildErr=!__MsbuildErr! %__RunArgs% %__UnprocessedBuildArgs%
+
```
Comment (tannergooding) [inline build.cmd]: nit: unnecessary whitespace

Comment (4creators) [reply build.cmd]: applied spacing convention from native build part where between log variables and build call is a space introduced intentionally  for clarity of code. Should I fix it in both places or leave them as they are now?

Comment (tannergooding) [reply build.cmd]: the newline is fine, I was pointing out the four whitespaces on the (otherwise empty) line 😄

Comment (4creators) [reply build.cmd]: Yeah, that was it 😄

**Comment by 4creators** [comment]
PTAL

All feedback addressed - I understand that @weshaggard comment regarding C++ team code review makes merging it in current shape possible.

**Comment by danmosemsft** [comment]
I think this fine (I worked on the C++ tasks long ago.. for what it's worth, I can't remember much..) . Thanks @4creators

**Comment by danmosemsft** [comment]
would be great if you want to look at build-tests. It's very slow.

**Comment by weshaggard** [comment]
LOL @danmosemsft I just read your old blog post on this topic https://blogs.msdn.microsoft.com/visualstudio/2010/03/07/tuning-c-build-parallelism-in-vs2010/ :)

Thanks @4creators I'm fine with this change although I kind of want to not pass in the props file to every build command and so I might try and figure out how to make Cmake automatically include it without requiring that.

**Comment by danmosemsft** [comment]
@weshaggard ha, I vaguely remember writing that ..

**Comment by 4creators** [comment]
>would be great if you want to look at build-tests. It's very slow.

Will do it - it's on my radar already

**Comment by 4creators** [comment]
>I kind of want to not pass in the props file to every build command

@weshaggard It is understandable, however with current solution it is possible to override all parameters set in script using additional command line arguments in call to build i.e.:
```bash
build ...... -- /p:ForceImportBeforeCppTargets=%__ProjectDir%/someotherpropsfile.props /m:30
```
This is why I have inserted `-ExtraParameters` before `%__UnprocessedBuildArgs%` in call to build cmd.

**Comment by BruceForstall** [comment]
One problem with this change is that it hijacks `-ExtraParameters` for use by the build script itself. This means users of build.cmd can't themselves pass any extra parameters, because you now get:
```
Error: Parameter ExtraParameters specified more than once.
```

A new mechanism needs to be devised here that frees up `-ExtraParameters` to only be used by invokers of build.cmd, not by the script itself.

**Comment by 4creators** [comment]
@BruceForstall 
Yeah, good find. Will work on solution to this problem.

**Comment by BruceForstall** [comment]
If there's no good general solution, a "hack" would be to add a section to config.json that duplicates ExtraParameters with some new name, e.g., "-SpecialPassThroughUsedByBuild", just to leave ExtraParameters unused.
